lib/elixir/src/elixir_erl_pass.erl

```diff
@@ -242,11 +243,22 @@ translate({{'.', _, [Left, Right]}, Meta, []}, _Ann, S) when is_tuple(Left), is_
     {clause, Generated,
       [TVar],
       [[?remote(Generated, erlang, is_map, [TVar])]],
-      [?remote(Ann, erlang, error, [TError])]},
+      [?remote(Ann, erlang, error, [TBadKeyError])]},
+    {clause, Generated,
```

Comment:
I'm creating more clauses directly in the generated code: should I be delegating to some `dot_apply` function instead? (that would be `@doc false` maybe?)

Reply:
We definitely want to avoid generating additional code and we want to avoid changing the semantics of the failure. My suggestion is the following:

1. If no_parens is true, we generate the existing bad key error (and we can improve the BadKeyError message as a whole)

2. If no_parens is false, we raise the undefined function error (and we can improve the UndefinedFunctionError as a whole)

I know there will still be some ambiguity in relation to apply or Map.fetch(nil, ...) but I think it is important to not add new exception types.

In the future, we can use the error_info from EEP54 to add some special annotation (and annotate that we were indeed in a dot or a similar).

Reply:
> In the future, we can use the error_info from EEP54 to add some special annotation (and annotate that we were indeed in a dot or a similar).

Oh, I see, this is the piece I was missing to be able to propagate this information without making a mess. Will wait for EEP54 for the dot-aware messages.

> we want to avoid changing the semantics of the failure

Makes perfect sense, will try the approach you suggested 👍

Reply:
EEP54 requires OTP 24, so we should add that particular bit later.